python/semantic_kernel/connectors/ai/open_ai/models/chat/azure_chat_with_data_response.py

```diff
@@ -0,0 +1,72 @@
+# Copyright (c) Microsoft. All rights reserved.
+"""Azure OpenAI Chat With Data Streaming Response class."""
+from typing import Tuple
+from openai import AsyncStream
+from openai.types.chat import ChatCompletionChunk
+from semantic_kernel.connectors.ai.chat_request_settings import ChatRequestSettings
+
+
+class AzureChatWithDataStreamResponse:
+    """Class to hold Azure OpenAI Chat With Data streaming response."""
+
+    _tool_message: str
+    _assistant_message: AsyncStream[ChatCompletionChunk]
+    _settings: "ChatRequestSettings"
+    
+    def __init__(self, assistant_message: AsyncStream[ChatCompletionChunk], settings: "ChatRequestSettings"):
+        self._assistant_message = assistant_message
+        self._tool_message = ""
+        self._settings = settings
+
+    async def _iterate_to_assistant_message(self):
+        """Iterate through the message stream to populate the tool message and prompt filter results and stop at the assistant message."""
+        while True:
+            try:
+                message = await self._assistant_message.__anext__()
+
+                if message.choices and len(message.choices) > 0:
+                    delta = message.choices[0].delta
+                    if delta:
+                        if delta.model_extra and "context" in delta.model_extra:
+                            for m in delta.model_extra["context"].get("messages", []):
+                                if m["role"] == "tool":
+                                    self._tool_message = m.get("content", "")
+                                    break
+                            break
+            except StopIteration:
+                break
+
+    async def get_tool_message(self):
+        """Get the tool message."""
+        if not self._tool_message:
+            await self._iterate_to_assistant_message()
+
+        return self._tool_message
+
+    @staticmethod
+    def _parse_choices(choice) -> Tuple[str, int]:
+        message = ""
+        if choice.delta.content:
+            message += choice.delta.content
+
+        return message, choice.index
+
+    async def __aiter__(self):
+        """Iterate through the message stream and yield the assistant response."""
+        if not self._tool_message:
+            await self._iterate_to_assistant_message()
+
+        async for chunk in self._assistant_message:
```

Comment:
Similar to a previous comment, we have this code in `async def complete_chat_stream_async` as part of `OpenAIChatCompletionBase`.

Reply:
In this case, the stream returned from the extensions chat completions API isn't just the assistant message deltas - the first message will be the "tool" message, then a message indicating role "assistant", and then the assistant message deltas start from the next message. So this class is intended to wrap the stream so that the first few events are processed to get out the tool message, and then the assistant message deltas can be consumed by the client in an `async for`. If we use the base class there's no way for the client to access the tool message.